tests/test_datasets.py

```
@@ -0,0 +1,101 @@
+"""Test core Datasets and DataLoaders."""
+
+from typing import Tuple
+
+import numpy as np
+import pytest
+
+from constants import EXTRA_VARS, INPT_VARS, OUT_VARS
+from datasets import TrainData
+from train_3D import Trainer
+
+# Note: Refactoring data loaders is planned for the near-term. Ideally,
+# these fixtures allow us to isolate data loader tests from their setup.
+
+
+# This micro-fixture is cached by pytest. Thus, we don't have to change
+# the factory methods that throw errors during double initialization.
```

What factory methods that throw errors is this referring to?

---

Boy, there are a lot! These are all throughout the codebase. On first glance, they look like singletons, but instead of returning the singleton instance, they throw an error. 

e.g. see the first few results of `_instance`:
https://github.com/search?q=repo%3Asuryadheeshjith%2FOcean_Emulator%20_instance&type=code